gravitee-rest-api-services/gravitee-rest-api-services-subscription-pre-expiration-notif/src/main/java/io/gravitee/rest/api/services/subscriptionpreexpirationnotif/ScheduledSubscriptionPreExpirationNotifService.java

```diff
@@ -0,0 +1,183 @@
+import org.springframework.scheduling.support.CronTrigger;
+
+public class ScheduledSubscriptionPreExpirationNotifService extends AbstractService implements Runnable {
+
+    private final Logger logger = LoggerFactory.getLogger(ScheduledSubscriptionPreExpirationNotifService.class);
+
+    @Autowired
+    private TaskScheduler scheduler;
+
+    @Value("${* * */1 * * *}")
+    private String cronTrigger;
+
+    private int cronPeriodInMs = 60 * 60 * 1000;
+
+    @Value("#{'${services.subscription.preExpirationNotificationSchedule:90,45,30}'.split(',')}")
+    private List<Integer> configPreExpirationNotificationSchedule;
+
+    private List<Integer> notificationDays;
+
+    private final AtomicLong counter = new AtomicLong(0);
+
+    @Autowired
+    private SubscriptionService subscriptionService;
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private RoleService roleService;
+
+    @Autowired
+    private MembershipService membershipService;
+
+    @Autowired
+    private EmailService emailService;
+
+    @Autowired
+    private ApiService apiService;
+
+    @Autowired
+    private PlanService planService;
+
+    @Autowired
+    private ApplicationService applicationService;
+
+    @Override
+    protected String name() {
+        return "Subscription Pre Expiration Notification service";
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        super.doStart();
+
+        notificationDays = getCleanedNotificationDays(configPreExpirationNotificationSchedule);
+
+        logger.info("Subscription Pre Expiration Notification service has been initialized with cron [{}]", cronTrigger);
+        scheduler.schedule(this, new CronTrigger(cronTrigger));
+    }
+
+    @Override
+    public void run() {
+        logger.debug("Subscription Pre Expiration Notification #{} started at {}", counter.incrementAndGet(), Instant.now().toString());
+
+        Date now = new Date();
+
+        notificationDays.forEach(
+            preExpirationDuration ->
+                findSubscriptionToNotify(now, preExpirationDuration)
+                    .forEach(
+                        subscriptionEntity -> {
+                            ApiEntity api = apiService.findById(subscriptionEntity.getApi());
+                            PlanEntity plan = planService.findById(subscriptionEntity.getPlan());
+
+                            ApplicationEntity application = applicationService.findById(subscriptionEntity.getApplication());
+
+                            findEmailsToNotify(subscriptionEntity)
+                                .forEach(email -> this.sendEmail(email, preExpirationDuration, api, plan, application));
+                        }
+                    )
+        );
+
+        logger.debug("Subscription Pre Expiration Notification #{} ended at {}", counter.get(), Instant.now().toString());
+    }
+
+    protected List<Integer> getCleanedNotificationDays(List<Integer> inputDays) {
```

Comment:
is `protected` only for testing purpose ? in this case, maybe add a `@VisibleForTesrting` ?